File: src/ros1_for_reference/dofbot_moveit/src/03_attached_object.cpp
#include <iostream>
#include "ros/ros.h"
#include <moveit/move_group_interface/move_group_interface.h>
#include <tf/LinearMath/Quaternion.h>
#include <moveit_visual_tools/moveit_visual_tools.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>

using namespace std;

int main(int argc, char **argv) {
    ros::init(argc, argv, "dofbot_attached_object_cpp");
    ros::NodeHandle n;
    // 设置线程
    ros::AsyncSpinner spinner(1);
    // 开启线程
    spinner.start();
    // Initialize the robotic arm motion planning group
    // 初始化机械臂运动规划组
    moveit::planning_interface::MoveGroupInterface dofbot("dofbot");
    dofbot.allowReplanning(true);
    // (单位：秒)  (unit: second)
    dofbot.setPlanningTime(5);
    dofbot.setNumPlanningAttempts(10);
    // (单位：米)  (unit: m)
    dofbot.setGoalPositionTolerance(0.01);
    // (单位：弧度)  (unit: radian)
    dofbot.setGoalOrientationTolerance(0.01);
    // 设置最大速度   set maximum speed
    dofbot.setMaxVelocityScalingFactor(1.0);
    // 设置最大加速度  set maximum acceleration
    dofbot.setMaxAccelerationScalingFactor(1.0);
    //设置目标点  set target point
    dofbot.setNamedTarget("up");
    //开始移动  start moving
    dofbot.move();
    sleep(0.1);
    // Get the name of the framework the bot is planning
    // 获取机器人正在计划的框架的名称
    string frame = dofbot.getPlanningFrame();
    // 创建场景实例  Create scene instance
    moveit::planning_interface::PlanningSceneInterface scene;
    ////////////// 添加障碍物 Add obstacles ////////////////
    vector<string> object_ids;
    scene.removeCollisionObjects(object_ids);
    // 创建检测对象容器  Create a detection object container
    vector<moveit_msgs::CollisionObject> objects;
    // 创建碰撞检测对象  Create Collision Detection Objects
    moveit_msgs::CollisionObject obj;
    // 设置障碍物的id  Set the id of the obstacle
    obj.id = "obj";
    object_ids.push_back(obj.id);
    // 障碍物的状态  the status of the obstacle
    obj.operation = obj.ADD;
    // 设置障碍物的头信息  Set the header information of the obstacle
    obj.header.frame_id = frame;
    shape_msgs::SolidPrimitive primitive;
    // 设置障碍物类型  Set the obstacle type
    primitive.type = primitive.BOX;
    // 设置障碍物维度  Set Obstacle Dimensions
    primitive.dimensions.resize(3);
    // 设置障碍物的长宽高  Set the length, width and height of obstacles
    primitive.dimensions[0] = 0.1;
    primitive.dimensions[1] = 0.2;
    primitive.dimensions[2] = 0.1;
    obj.primitives.push_back(primitive);
    geometry_msgs::Pose pose;
    // 设置障碍物的位置信息[x,y,z]  Set the position information of the obstacle [x,y,z]
    pose.position.x = 0;
    pose.position.y = 0.2;
    pose.position.z = 0.3;
    // 设置障碍物的姿态信息  Set the attitude information of obstacles
    tf::Quaternion quaternion;
    // R,P,Y的单位是角度  The units of R,P,Y are degrees
    double Roll = 0.0;
    double Pitch = 0.0;
    double Yaw = 90.0;
    // RPY转四元数  RPY to Quaternion
    quaternion.setRPY(Roll * M_PI / 180, Pitch * M_PI / 180, Yaw * M_PI / 180);
    pose.orientation.x = quaternion.x();
    pose.orientation.y = quaternion.y();
    pose.orientation.z = quaternion.z();
    pose.orientation.w = quaternion.w();
    // 设置障碍物的位姿信息 Set the pose information of obstacles
    obj.primitive_poses.push_back(pose);
    objects.push_back(obj);
    /////////// 设置障碍物的颜色 Set the color of the obstacle ////////////////////
    // 创建检测对象的颜色容器 Create a color container for detected objects
    std::vector<moveit_msgs::ObjectColor> colors;
    // 创建颜色实例 Create a color instance
    moveit_msgs::ObjectColor color;
    // 添加需要设置颜色的id Add the id that needs to set the color
    color.id = "obj";
    // 设置RGBA值,范围[0~1] Set RGBA value, range [0~1]
    color.color.r = 0;
    color.color.g = 1.0;
    color.color.b = 0;
    color.color.a = 0.5;
    colors.push_back(color);
    // 将设置的信息添加到场景中 Add set information to the scene
    scene.applyCollisionObjects(objects, colors);
    // 设置目标点 set target point
    dofbot.setNamedTarget("down");
    // 开始移动 start moving
    dofbot.move();
    sleep(0.5);
    while (true){
        // 设置随机目标点 Set random target points
        dofbot.setRandomTarget();
        // 开始移动 start moving
        dofbot.move();
        sleep(0.5);
    }
    // 阻塞进程 blocking process
    ros::waitForShutdown();
    return 0;
}

File: src/ros1_for_reference/dofbot_moveit/scripts/03_attached_object.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import rospy, sys
from time import sleep
import moveit_commander
from geometry_msgs.msg import PoseStamped
from moveit_commander import MoveGroupCommander, PlanningSceneInterface

if __name__ == "__main__":
    moveit_commander.roscpp_initialize(sys.argv)
    rospy.init_node('dofbot_attached_object_py')
    # Initialize the scene object
    # 初始化场景对象
    scene = PlanningSceneInterface()
    # Initialize the robotic arm motion planning group
    # 初始化机械臂运动规划组
    dofbot = MoveGroupCommander("dofbot")
    # Allow replanning when motion planning fails
    # 当运动规划失败后，允许重新规划
    dofbot.allow_replanning(True)
    dofbot.set_planning_time(5)
    # number of attempts to plan
    # 尝试规划的次数
    dofbot.set_num_planning_attempts(10)
    # Set allowable target position error
    # 设置允许目标位置误差
    dofbot.set_goal_position_tolerance(0.01)
    # Set the allowable target attitude error
    # 设置允许目标姿态误差
    dofbot.set_goal_orientation_tolerance(0.01)
    # Set allowable target error
    # 设置允许目标误差
    dofbot.set_goal_tolerance(0.01)
    # set maximum speed
    # 设置最大速度
    dofbot.set_max_velocity_scaling_factor(1.0)
    # set maximum acceleration
    # 设置最大加速度
    dofbot.set_max_acceleration_scaling_factor(1.0)
    # Set "up" as the target point
    # 设置"up"为目标点
    dofbot.set_named_target("up")
    dofbot.go()
    sleep(0.5)
    # Set the height of the desktop
    # 设置桌面的高度
    table_ground = 0.2
    # Set the 3D size of the obstacle [length, width and height]
    # 置障碍物的三维尺寸[长宽高]
    table_size = [0.7, 0.1, 0.02]
    # Add the table to the scene
    # 将table加入场景当中
    table_pose = PoseStamped()
    table_pose.header.frame_id = 'base_link'
    table_pose.pose.position.x = 0
    table_pose.pose.position.y = 0.15
    table_pose.pose.position.z = table_ground + table_size[2] / 2.0
    table_pose.pose.orientation.w = 1.0
    scene.add_box('table', table_pose, table_size)
    rospy.sleep(2)
    dofbot.set_named_target("down")
    dofbot.go()
    sleep(0.5)
    moveit_commander.roscpp_shutdown()
    moveit_commander.os._exit(0)

